controllers/admin_controller.rb

```diff
@@ -0,0 +1,168 @@
+    @atm_view = AtmView.new
+    @validator = Validator.instance
+  end
+
+  def run
+    loop do
+      @view.print_header("Bank Admin CLI")
+      @view.print_menu([
+        "Create Account",
+        "Create ATM",
+        "List Accounts",
+        "List Transactions",
+        "List ATMs",
+        "Edit Account",
+        "Edit ATM",
+        "Exit"
+      ])
+      choice = @view.prompt("Choose an option")
+      case choice
+      when '1' then create_account
+      when '2' then create_atm
+      when '3' then list_accounts
+      when '4' then list_transactions
+      when '5' then list_atms
+      when '6' then edit_account
+      when '7' then edit_atm
+      when '8' then break
+      else @view.print_invalid_option
+      end
+    end
+  end
+
+  private
+
+  def create_account
+    @view.print_header("Create New Bank Account")
+    params = {
+      name:    @view.prompt("Full Name"),
+      job:     @view.prompt("Job Title"),
+      email:   @view.prompt("Email Address"),
+      address: @view.prompt("Full Address")
+    }
+    invalid_fields = params.keys.reject { |field| @validator.valid_field(field, params[field]) }
+    unless invalid_fields.empty?
+      invalid_fields.each { |field| @view.print_invalid_field(field) }
+      return
+    end
+    result = @account_service.create_account(**params)
+    if result[:error]
+      @account_view.creation_failure(result[:error])
+    else
+      @account_view.creation_success(result)
+    end
+  end
+
+  def create_atm
+    @view.print_header("Add New ATM")
+    location = @view.prompt("ATM Location")
+    unless @validator.valid_location?(location)
+      @atm_view.creation_failure("Invalid location")
+      return
+    end
+    atm = @atm_service.create_atm(location)
+    if atm.is_a?(Hash) && atm[:error]
+      @atm_view.creation_failure(atm[:error])
+    else
+      @atm_view.creation_success(atm)
+    end
+  end
+
+  def list_accounts
+    accounts = @account_service.get_all_accounts
+    @account_view.list_accounts(accounts)
+  end
+
+  def list_transactions
+    @view.print_header("List Transactions")
+    account_id = @view.prompt("Enter Account ID")
+
+    unless @validator.valid_id?(account_id)
+      @view.print_error("Invalid Account ID.")
+      return
+    end
+
+    account_id = account_id.to_i
+    result = @account_service.get_account(account_id)
+    return @view.print_error(result[:error]) if result[:error]
+
+    account = result[:account]
+    transactions = @transaction_service.get_transactions_for_account(account.id)
+    @transaction_view.list_transactions(account, transactions)
+  end
+
+  def list_atms
+    atms = @atm_service.get_all_atms
+    @atm_view.list_atms(atms)
+  end
+
+  def edit_account
+    @view.print_header("Edit Account - blank fields are not changed")
+    account_id = @view.prompt("Enter Account ID to edit")
+    unless @validator.valid_id?(account_id)
+      @view.print_error("Invalid Account ID.")
+      return
+    end
```

Comment:
Not necessarily for this assignment, but it would clean up the code quite a bit if you would throw an error in the service and have catching mechanism that's printing the message.